README.md
# embed-parser

[![Build Status][travis-image]][travis-url]
[![Coverage Status][codecov-image]][codecov-url]
[![NPM Downloads][downloads-image]][downloads-url]
[![NPM Version][version-image]][version-url]
[![License][license-image]][license-url]
[![Dependency Status][dependency-image]][dependency-url]
[![devDependency Status][devdependency-image]][devdependency-url]
[![Code Style][style-image]][style-url]

> A parser for get embed data from url, use [oembed-parser](https://www.npmjs.com/package/oembed-parser) for fallback

## Installation

```shell
$ yarn add embed-parser

# or npm
$ npm install embed-parser
```

## Usage

```javascript
const embedParser = require('embed-parser')

embedParser('https://v.youku.com/v_show/id_XMzkyODgxODM2OA==.html').then(embed => {
  console.log(embed)
  // =>
  //  {
  //    type: 'video',
  //    version: '1.0',
  //    title: 'YouKu Video',
  //    provider_name: 'YouKu',
  //    provider_url: 'https://www.youku.com',
  //    html: '<iframe src="http://player.youku.com/embed/XMzkyODgxODM2OA==" width="640" height="360" frameborder="0" allowfullscreen></iframe>',
  //    width: 640,
  //    height: 360
  //  }
})
```

## Features

- Extract embed by local [rules](lib/handler/local-rules.json)
- use [OEmbed](https://oembed.com) for fallback
- [Custom embed handler](#example)

## API

### embedParser(url)

Extract embed data from url.

#### url

- Type: `string`
- Details: original url

### embedParser.extract(url)

- alias to `embedParser(url)`

### embedParser.hasHandler(url)

Determine if there is a matching handler.

#### url

- Type: `string`
- Details: original url

### embedParser.hasProvider(url)

- alias to `embedParser.hasHandler(url)`

### embedParser.registerHandler(regex, handler, priority = 20)

Register an custom embed handler.

#### regex

- Type: `string`
- Details: The regex that will be used to see if this handler should be used for a URL.

#### handler

- Type: `function`
- Details: The handler function that will be called if the regex is matched.

#### priority

- Type: `number`
- Details: The handler function priority
- Default: 20

#### example

```javascript
embedParser.registerHandler('http://v.zce.me/item-(.+)', matches => {
  return { html: `http://v.zce.me/embed/${matches[1]}` }
})
```

## Contributing

1. **Fork** it on GitHub!
2. **Clone** the fork to your own machine.
3. **Checkout** your feature branch: `git checkout -b my-awesome-feature`
4. **Commit** your changes to your own branch: `git commit -am 'Add some feature'`
5. **Push** your work back up to your fork: `git push -u origin my-awesome-feature`
6. Submit a **Pull Request** so that we can review your changes.

> **NOTE**: Be sure to merge the latest from "upstream" before making a pull request!

## License

[MIT](LICENSE) &copy; [汪磊](https://zce.me)



[travis-image]: https://img.shields.io/travis/zce/embed-parser.svg
[travis-url]: https://travis-ci.org/zce/embed-parser
[codecov-image]: https://img.shields.io/codecov/c/github/zce/embed-parser.svg
[codecov-url]: https://codecov.io/gh/zce/embed-parser
[downloads-image]: https://img.shields.io/npm/dm/embed-parser.svg
[downloads-url]: https://npmjs.org/package/embed-parser
[version-image]: https://img.shields.io/npm/v/embed-parser.svg
[version-url]: https://npmjs.org/package/embed-parser
[license-image]: https://img.shields.io/npm/l/embed-parser.svg
[license-url]: https://github.com/zce/embed-parser/blob/master/LICENSE
[dependency-image]: https://img.shields.io/david/zce/embed-parser.svg
[dependency-url]: https://david-dm.org/zce/embed-parser
[devdependency-image]: https://img.shields.io/david/dev/zce/embed-parser.svg
[devdependency-url]: https://david-dm.org/zce/embed-parser?type=dev
[style-image]: https://img.shields.io/badge/code_style-standard-brightgreen.svg
[style-url]: http://standardjs.com


CHANGELOG.md
# Changelog

## [0.1.0] - 2018-11-23

- Initial release

<!-- http://keepachangelog.com/ -->


test/embed-parser.test.js
import test from 'ava'
import embedParser from '..'
import { version } from '../package'

test('extract-local-youku', async t => {
  const embed = await embedParser('https://v.youku.com/v_show/id_XMzkyODgxODM2OA==.html')
  t.is(embed.provider_name, 'YouKu')
  t.is(embed.html, '<iframe src="http://player.youku.com/embed/XMzkyODgxODM2OA==" width="640" height="360" frameborder="0" allowfullscreen></iframe>')
})

test('extract-local-bilibili', async t => {
  const embed = await embedParser('https://www.bilibili.com/video/av21536312')
  t.is(embed.provider_name, 'Bilibili')
  t.is(embed.html, '<iframe src="http://player.bilibili.com/player.html?aid=21536312" width="640" height="360" frameborder="0" allowfullscreen></iframe>')
})

test('extract-oembed', async t => {
  const embed = await embedParser('https://codepen.io/zce/pen/JepwwN')
  t.is(embed.provider_name, 'Codepen')
})

test('extract-custom', async t => {
  embedParser.registerHandler('http://v.zce.me/item-(.+)', matches => {
    return Promise.resolve({ url: `http://v.zce.me/embed/${matches[1]}` })
  })
  const embed = await embedParser('http://v.zce.me/item-fake-slug')
  t.is(embed.url, 'http://v.zce.me/embed/fake-slug')
})

test('extract-failed', async t => {
  await t.throwsAsync(
    () => embedParser('https://fakkkkkkkkkkkkker.com/item-fake-slug'),
    'No handler found with given url "https://fakkkkkkkkkkkkker.com/item-fake-slug"'
  )
})

test('extract-with-invalid-url', async t => {
  await t.throwsAsync(() => embedParser('invalid url'), 'Invalid input URL')
  await t.throwsAsync(() => embedParser(), 'Invalid input URL')
})

test('hasHandler-local', t => {
  t.true(embedParser.hasHandler('https://v.youku.com/v_show/id_XMzkyODgxODM2OA==.html'))
  t.false(embedParser.hasHandler('https://fakkkkkkkkkkkkker.com'))
})

test('hasHandler-oembed', t => {
  t.true(embedParser.hasHandler('https://codepen.io/zce/pen/JepwwN'))
  t.false(embedParser.hasHandler('https://fakkkkkkkkkkkkker.com'))
})

test('hasHandler-custom', t => {
  embedParser.registerHandler('http://v.zce.me/item-(.+)', () => {})
  t.true(embedParser.hasHandler('http://v.zce.me/item-fake-slug'))
})

test('hasHandler-with-invalid-url', t => {
  t.throws(() => embedParser.hasHandler('invalid url'), 'Invalid input URL')
  t.throws(() => embedParser.hasHandler(), 'Invalid input URL')
})

test('registerHandler', async t => {
  const fn = () => {}
  const addedHandler = embedParser.registerHandler('\\w+', fn, 50).shift()
  t.true(addedHandler.regex.test('zce'))
  t.is(addedHandler.handler, fn)
})

test('compatible-oembed-parser', t => {
  // extract
  t.is(embedParser.extract, embedParser)
  // hasProvider
  t.is(embedParser.hasProvider, embedParser.hasHandler)
  // version
  t.is(embedParser.version, version)
})


lib/index.js
const handler = require('./handler')
const { version } = require('../package')

const isValidURL = input => {
  if (!input) return false
  const regex = /^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,}))\.?)(?::\d{2,5})?(?:[/?#]\S*)?$/i
  return regex.test(input)
}

/**
 * Extract embed data from url.
 * @param {String} url Url.
 * @returns {Promise<Object>} A promise to return embed data.
 */
const embedParser = url => {
  // validate url
  if (!isValidURL(url)) { return Promise.reject(new Error('Invalid input URL')) }

  return handler.extract(url)
}

/**
 * Determine if there is a matching handler.
 * @param {String} url Url.
 * @returns {Boolean} Yes or no.
 */
embedParser.hasHandler = url => {
  // validate url
  if (!isValidURL(url)) throw new Error('Invalid input URL')

  return !!handler.find(url)
}

/**
 * Register an custom embed handler.
 * @param {String} regex The regex that will be used to see if this handler should be used for a URL.
 * @param {Function} handler The handler function that will be called if the regex is matched.
 * @param {Number} priority The handler function priority, default: 20.
 * @returns {Array} An array containing all the handlers.
 */
embedParser.registerHandler = handler.register

// Compatible with oembed-parser api
embedParser.version = version
embedParser.extract = embedParser
embedParser.hasProvider = embedParser.hasHandler

module.exports = embedParser


lib/handler/local.js
/**
 * Local embed handler
 */

const BaseHandler = require('./base')
const rules = require('./local-rules')

/**
 * Compile an ES6 template literals to a Template function
 * @param {String} source ES6 template literals
 */
const compile = source => {
  return context => {
    /* eslint-disable no-new-func */
    const props = Object.keys(context).join(', ')
    return new Function(`{ ${props} }`, `return \`${source}\``)(context)
  }
}

const getEmbed = (rule, matches) => {
  const embed = Object.assign({}, rule)
  delete embed.regex

  // https://oembed.com/#section2.3

  const context = Object.assign({ matches }, rule)
  const isTemplate = input => /\${[^{}]+}/.test(input)

  Object.keys(embed).forEach(key => {
    const item = embed[key]
    if (typeof item === 'string' && isTemplate(item)) {
      embed[key] = compile(embed[key])(context)
    }
  })

  // remove regex
  return embed
}

module.exports = class LocalHandler extends BaseHandler {
  constructor () {
    super()
    this.rules = rules.map(item => {
      if (!item.regex) throw new Error('`local-rule.js` is invalid: Must have regex attribute')
      item.regex = new RegExp(item.regex)
      return item
    })
  }

  match (url) {
    return this.rules.some(item => item.regex.test(url))
  }

  extract (url) {
    const rule = this.rules.find(item => item.regex.test(url))
    // try to extract matches from url
    const matches = rule.regex.exec(url)
    if (!matches) return Promise.reject(new Error(`Can not extract embed from given url "${url}"`))

    // embed data
    const embed = getEmbed(rule, matches)
    return Promise.resolve(embed)
  }
}


example/usage.js
const embedParser = require('..')

embedParser.hasHandler('https://v.youku.com/v_show/id_XMzkyODgxODM2OA==.html')
// => true

embedParser.hasHandler('https://www.bilibili.com/video/av21536312')
// => true

embedParser('https://v.youku.com/v_show/id_XMzkyODgxODM2OA==.html').then(embed => {
  console.log(embed)
  // =>
  //  {
  //    type: 'video',
  //    version: '1.0',
  //    title: 'YouKu Video',
  //    provider_name: 'YouKu',
  //    provider_url: 'https://www.youku.com',
  //    html: '<iframe src="http://player.youku.com/embed/XMzkyODgxODM2OA==" width="640" height="360" frameborder="0" allowfullscreen></iframe>',
  //    width: 640,
  //    height: 360
  //  }
})

embedParser('https://www.bilibili.com/video/av21536312').then(embed => {
  console.log(embed)
  // =>
  //  {
  //    type: 'video',
  //    version: '1.0',
  //    title: 'Bilibili Video',
  //    provider_name: 'Bilibili',
  //    provider_url: 'https://www.bilibili.com,
  //    html: '<iframe src="http://player.bilibili.com/player.html?aid=21536312" width="640" height="360" frameborder="0" allowfullscreen></iframe>',
  //    width: 640,
  //    height: 360
  //  }
})

embedParser.registerHandler('http://v.zce.me/item-(.+)', matches => new Promise(resolve => {
  resolve({ url: `http://v.zce.me/embed/${matches[1]}` })
}))
embedParser('http://v.zce.me/item-fake-slug').then(embed => {
  console.log(embed)
  // =>
  //  {
  //    url: 'http://v.zce.me/embed/fake-slug'
  //  }
})
